nwb_conversion_tools/utils/json_schema.py

```diff
@@ -31,33 +31,50 @@ def append_replace_dict_in_list(d, ls, k):
     """
     if k in d and len(ls) > 0:
         # Index where the value dictionary[k] exists in the list of dicts
-        ind = np.where([d[k] == i[k] for i in ls])[0]
-        if len(ind) > 0:
-            ls[ind[0]] = d
+        indxs = np.where([d[k] == i[k] for i in ls])[0]
+        if len(indxs) > 0:
+            for idx in indxs:
+                ls[idx] = dict_deep_update(ls[idx], d)
         else:
             ls.append(d)
     else:
         ls.append(d)
     return ls
 
 
-def dict_deep_update(d: dict, u: dict, append_list: bool = True, remove_repeats: bool = True) -> dict:
+def dict_get_dtype(d: collections.abc.Mapping, key: str, default_val=None):
+    ret_val = d.get(key, default_val)
+    if isinstance(ret_val, type(default_val)) or default_val is None:
+        return ret_val
+    else:
+        return default_val
+
+
+def dict_deep_update(
+    d: collections.abc.Mapping,
+    u: collections.abc.Mapping,
+    append_list: bool = True,
+    remove_repeats: bool = True,
+    copy: bool = False,
+    compare_key: str = "name",
+) -> dict:
     """Perform an update to all nested keys of dictionary d from dictionary u."""
+    if copy:
+        d = deepcopy(d)
     for k, v in u.items():
         if isinstance(v, collections.abc.Mapping):
-            d[k] = dict_deep_update(d.get(k, {}), v, append_list=append_list, remove_repeats=remove_repeats)
+            d[k] = dict_deep_update(dict_get_dtype(d, k, {}), v, append_list=append_list, remove_repeats=remove_repeats)
         elif append_list and isinstance(v, list):
-            if len(v) > 0 and isinstance(v[0], dict):
-                for vv in v:
-                    d[k] = append_replace_dict_in_list(d=vv, ls=d.get(k, []), k="name")
-                    # add dict only if not repeated
-                    # if not exist_dict_in_list(vv, d.get(k, [])):
-                    # d[k] = d.get(k, []) + [vv]
-            else:
-                d[k] = d.get(k, []) + v
-                # Remove repeated items if they exist
-                if remove_repeats and len(d[k]) > 0:
-                    d[k] = list(set(d[k]))
+            d[k] = dict_get_dtype(d, k, [])
+            for vv in v:
```

Comment:
it loops over the elements of list and does not need all the elements to be of the same dtype. A more general case from before, where we assume all elements of the list would be dictionaries